### ISSUE TITLE
CI Infrastructure

### ISSUE THREAD

**Issue by chriscool** [ipfs/infrastructure#100]
As we investigate golang/build (see issue #99) we might want to figure out what would make us super happy both in the short and long term.

The basic goals are:
- run builds for various platforms
- run tests in various platforms
- run longer network/protocol tests on demand

And two constraints are:
- having a reasonable scalability path to dozens, hundreds, thousands, then even millions of nodes. with thousands, then millions of jobs; these might mean multiple tools along the way.
- untrusted / well-scoped trust environment, sandbox heavily, job redundancy or maybe a trust graph.

Related issues:
- https://github.com/ipfs/infrastructure/issues/25 (CI build monitoring)
- https://github.com/ipfs/infrastructure/issues/20 (Testing Infrastructure)

**Comment by chriscool** [comment]
Related issue: https://github.com/ipfs/infrastructure/issues/25 (CI build monitoring)

**Comment by diasdavid** [comment]
In today's sprint hangout, CI was brought up as one of the main points that are blocking our development teams from moving faster. I was going to open a issue in ipfs/community, but since this one already existed, let's continue here.

**Current issues:**

- circle-ci and travis-ci are slow and they don't let us upgrade our account, since all our projects are open source
- we get into 'travis queue' a lot, since the majority of the projects belong to the same github organization, we hit the CI runs limit faster, sometimes having to wait even 30 minutes just to start running the tests.

**Requirements:**

- Support several platforms (Linux, Mac, Windows)
- Support several browsers (Chrome, Firefox)
- Be blazing fast (we probably need auto scalling for when the team is working all at the same time)
- No queue waiting on CI
- Support custom environments (for interop tests)
- Support spawning thousands of nodes. If possible, distributed across different locations.

**Possible solutions:**

- TeamCity (cons: maintenance) 
- Buildkite (https://twitter.com/yoshuawuyts/status/764774088539865088)

@chriscool you have an incredible amount of experience testing different projects with very eccentric needs, if you feel like Captaining this endeavour and optimising our CI infrastructure to enable our dev team to move forward faster, either by leading or building, that would be amazing!

@yoshuawuyts, I know you are also really excited with CI and have been doing a lot of work with BuildKite, your thoughts on this are greatly appreciated.

Let's gather all ideas before Wednesday, August 17, so that we can have a call and commit to a plan to move forward. [Does 4:30pm UTC work for everyone](http://www.worldtimebuddy.com/?qm=1&lid=2267057,100,8,2950159&h=2267057&date=2016-8-15&sln=17.5-18.5)?

//cc @jbenet @whyrusleeping @flyingzumwalt @dignifiedquire

**Comment by jbenet** [comment]
- I _think_ Circle-CI does let us upgrade. verify this.
- I _think_ Travis-CI can be convinced. someone should contact them. (i'd offer to, but have other things)

**Comment by hermanjunge** [comment]
Can I be on this call?

**Comment by yoshuawuyts** [comment]
Hey @daviddias, thanks for pinging me. I won't be able to make the call (yay,
hosting a Nodeschool) so I'll just leave my reply here.

You're right in saying I'm quite a fan of buildkite. I've been using it for the
past week, and must say I'm impressed. For IPFS, some of the benefits I think
buildkite has over other implementations is:

- Precision management of nodes; you host your own workers and only pay
  buildkite for no of people in the org. You can also put workers on OSX,
  Windows, BSD etc. so cross-platform / device testing can become unified.
- Build configs are checked into source; this means that unlike some of the
  other self-hosted solutions there are no giant XML files to configure, but
  everything is nicely scoped.
- Zero knowledge; buildkite never checks out your code. They're used by banks.
  I think they're passing like these security checks and stuff. I think.
- and probably more stuff I'm forgetting

But there's one major drawback to all this: they don't have public build
support (yet). It's on the roadmap, but not a solution for right now.

So I've reached out to the @buildkite people today, asking them what's up with
public builds. Apparently other orgs such as Cordova use an (automated) invite
system that gets new people on the org whenever they make a PR. This apparently
works alright for them.

Another option @toolmantim suggested was creating a custom frontend for builds.
A custom page like `builds.ipfs.io` could work; though this has the drawback
that it would need to be implemented. Having a custom, public UI might be the
right solution in the long term, but given the current work load that might be
an unwelcome distraction to the timeline. The architecture he proposed was
this:
```txt
 ┌───────────────────────────┐           ┌────────────┐             ┌─────────────────────────┐
 │  github.com/ipfs/project  │           │ Buildkite  │             │ https://builds.ipfs.io  │
 └───────────────────────────┘           └────────────┘             └─────────────────────────┘
               │                                │                                │             
               │                                │                                │             
               ├───────────PR opened / ─────────▶                                │             
               │            fork push           │                                │             
               │             webhook            │                                │             
               │                                └───────────Events via───────────▶             
               │                                             webhooks            │             
               │                                                                 │             
               │                                                                 │             
               │                              POST                               │             
               ◀─────────────────────────────commit ─────────────────────────────┤             
               │                             status                              │             
               │                                                                 │             
               │                                                                 │             
               │                                                                 │             
               │                        Contributor can                          │             
               └──────────────────────click through or go ───────────────────────▶             
                                          directly to                                          
                                                                                               
                                                                                               
            Buildkite handles jobs dispatching, running agents, streaming logs,                
            capturing artifacts etc.                                                           
                                                                                               
                                                                                               
            builds.ipfs.io is a Node app that:                                                 
            - Receives Buildkite webhooks, posts commit status to GitHub                       
            - Display builds using Buildkite REST or GraphQL API (can use artifacts and        
            meta-data to instrument/customize the build results)                               
            - Serves up dynamic build badges based on Buildkite build info                     
```

Some of the things that were also discussed were full-public orgs; where every
single build is public info is public. As @toolmantim it might make sense to
run open source builds on a physically separate infrastructure from anything
that might contain keys / do deploys. So separating private / public might make
sense right there.

Ueh, and I think that's about it. I think if you holler at @toolmantim for
specific questions he can provide a better reply about the buildkite stuff than
I can.

Hope you all find a good solution. Cheers! :v:

### See Also
- [builds.ipfs.io architecture sketch](https://gist.github.com/toolmantim/20e008e6ff8ed75cd8f3beea14899d15)

**Comment by toolmantim** [comment]
👋 I just had a chat with @yoshuawuyts about it all. Just to clarify, nobody does the auto-invite thing just yet. But we're free for open-source, we just have to hit a button for you, and we're happy to support IPFS if you wanted to go w/ Buildkite. I'd recommend two completely separate Buildkite orgs in your case… one for third-party PRs, and another for core/internal use.

**Comment by hermanjunge** [comment]
Hi @toolmantim, I am evaluating a CI system for `ipfs-cluster`. Just to know: Do you provide support for docker? How about kubernetes?

(This is a little separate from the whole IPFS team goals, so take my questions with a grain of salt 😃 )

**Comment by yoshuawuyts** [comment]
@hermanjunge yeah it does. Have been using it for the past few days on google super tiny (micro?) nodes; and a few odd restarts aside (yay, low ram) they seem to be holding up quite well.

### See Also
- [article I wrote today about buildkite on k8](https://medium.com/@yoshuawuyts/yosh-in-infrastructure-land-36c54e1bb71f)
- [buildkite deployment yaml](https://github.com/yoshuawuyts/infrastructure/blob/master/buildkite/buildkite-agent-builder-deployment.yaml)

:v:

**Comment by toolmantim** [comment]
@hermanjunge yeah, the build agent runs on Docker and Kubernetes just fine, we just don't have the equivalent to https://github.com/buildkite/elastic-ci-stack-for-aws for Kubernetes just yet. But we have https://github.com/buildkite/helm-charts. And then for the actual build tasks themselves, they can run in Docker (for stable agent https://buildkite.com/docs/guides/docker-containerized-builds or for beta agents we've moved it into a plugin: https://github.com/buildkite-plugins/docker-compose-buildkite-plugin). Might want to move this to another issue to talk more tho?

**Comment by hermanjunge** [comment]
@toolmantim Amazing! I have some tests scenarios that involve building docker images, and deploying into kubernetes before running the tests. Will sign up, learn and provide feedback of Buildkite. Thank you very much for answering.

**Comment by whyrusleeping** [comment]
@toolmantim thanks for stopping by, buildkite looks really nice :)
for the record, i'm in love with teamcity, but i'm open to something that looks less enterprise-java-ey and still provides the same features.

It looks like with buildkite we still have to manually set up each repo we want CI for (same with teamcity). Are there any nice automations for setting up a 'stock CI setup for environment X'? Similar to how easy it is to get travis CI running?

teamcity provides really nice metrics such as build time per step, total tests run, number of tests passed and ignored. The latter few are because teamcity has plugins to parse and understand test output, does buildkite have anything similar?

Following along the previous point, We currently fail our tests if the number of tests executed drops significantly from the previous successful run. This is because teamcity allows us to define custom failure conditions based on both preset and user defined parameters. I think this is super cool and would hope that we have similar capabilities in any system we switch to.

It looks like buildkite supports windows, does it run scripts in cmd or git bash? (or can we specify?)

I notice in some of the documentation (that I admittedly havent read too far into) that there are buildkite hook files that get committed into vcs. It seems odd to have configuration both in the repo and on the buildkite admin page. Whats the purpose of that?

Anyways, I look forward to trying it out. I made an account and created an ipfs org, but haven't yet had time to make it any farther. Thanks!

**Comment by chriscool** [comment]
@diasdavid yeah I have some experience especially with Jenkins but it is not part of the possible solutions.
GitLab CI is also multi-platforms but maybe it is too much integrated with GitLab CE these days.
I will try to get more information about GitLab CI.
Also it will be difficult for me to attend tomorrow's call as I am having a vacation these days.

**Comment by diasdavid** [comment]
It seems to me that we have some solid action items before jumping onto a call, these are:

- [ ] Check if we are able to upgrade our circle-ci account being open source and what pricing options they have for our needs
- [ ] Reach out to travis-ci and learn if they have any plans of having more powerful open source plans
- [ ] Try buildkite -> Set it up for a go and a js repo and get a feeling of the experience of it
- [ ] Check what are the options with GitLab CI
- [ ] Evaluate Teamcity

Let's postpone the call until we have these items done, otherwise we might not be able to decide anything more than what we've discussed on this thread. It would be also good to have everyone when it happens.

**Comment by hermanjunge** [comment]
>  Try buildkite -> Set it up for a go and a js repo and get a feeling of the experience of it

@diasdavid I can commit to try buildkite, and do a quick video with a flow.

**Comment by whyrusleeping** [comment]
@hermanjunge If you want, i can invite you to the ipfs buildkite org and you can try setting something up there. It might make it easier for the rest of us to see

**Comment by hermanjunge** [comment]
@whyrusleeping Excellent. Do you have my email?

**Comment by whyrusleeping** [comment]
@hermanjunge yeap! sent out an invite

**Comment by chriscool** [comment]
About GitLab CI it looks like the best is to setup Repository Mirroring on GitLab: http://docs.gitlab.com/ee/workflow/repository_mirroring.html, as GitLab now integrates GitLab CI.
There is also the Quick Start docs for CI: http://docs.gitlab.com/ce/ci/quick_start/README.html
I will try to test this. Maybe I could test this on the Sharness repo.

**Comment by jbenet** [comment]
Can gitlab CI work as a github CI indicator? Could ease the transition

**Comment by jbenet** [comment]
(Of their potential users)

**Comment by chriscool** [comment]
@jbenet I think so. By the way you may want to read dak1 comments on this HN thread about yesterday's GitLab release: https://news.ycombinator.com/item?id=12338096

**Comment by hermanjunge** [comment]
Hi @toolmantim Our buildkite trial just expired and haven't had the chance to build a proper test yet. Can we get an extension trial, so I can describe my experience here ??? 😬 

Here is the email

<img width="1107" alt="screen shot 2016-08-29 at 11 12 47 pm" src="https://cloud.githubusercontent.com/assets/729830/18073699/4fe9302a-6e3e-11e6-9bb0-f54a9301af22.png">

(cc @whyrusleeping @yoshuawuyts )

**Comment by toolmantim** [comment]
@hermanjunge all fixed, I just extended it for a month. If this Buildkite org is for all third-party and public builds, I'll switch it to a free/open-source account if you like. But I'd suggest a separate org for private/sensitive/secure build pipelines, for doing all the releases etc, with a completely separate set of build agents (in their own VPC etc). `ipfs-core` for example. Anywho, fell free to bug me if you need a hand.

**Comment by yoshuawuyts** [comment]
Stumbled upon [this article](https://www.blackhat.com/docs/eu-15/materials/eu-15-Mittal-Continuous-Intrusion-Why-CI-Tools-Are-An-Attackers-Best-Friend.pdf) about popular self-hosted CI exploits; figured it might be relevant to the topics discussed here

**Comment by backus** [comment]
Hey guys. The buildkite team linked me to this issue. If you would like to take a look I just implemented buildkite for an open source project I work on in my free time. See this PR: https://github.com/backus/rubocop-rspec/pull/205

**Comment by whyrusleeping** [comment]
@backus thats awesome :)

Any idea how you would go about doing tests on windows and/or OSX?

**Comment by backus** [comment]
@whyrusleeping that I can't comment on if you need the OS inside the docker container to be windows or OS X.

**Comment by hermanjunge** [comment]
@backus The main idea of using containers is that you can be OS agnostic ;-)

**Comment by backus** [comment]
@hermanjunge yeah I get that. The container I use in the linked PR is just alpine linux. I figured that @whyrusleeping was asking about testing if some software works on different operating systems or something and I don't know about that.

**Comment by jbenet** [comment]
Woah, hey @backus o/ 😄

**Comment by backus** [comment]
:wave: Hey @jbenet

**Comment by VictorBjelkholm** [comment]
I've been investigating BuildKite a bit, trying to build, test and deploy projects via two agents I
run locally on two different machines, osx and ubuntu. Everything is working fine now for the five different projects I've tried.

How BuildKite fits the basic goals

* [X] run builds for various platforms

Easy to spin up docker containers, customized for the environment (go-buildkite, js-buildkite and so on),
or performing docker-in-docker builds (for building image, testing, tagging and deploying it)

Agents run on most OSes out there as well, https://buildkite.com/docs/agent/installation
		
* [X] run tests in various platforms

Up to us to setup the testlines, would be good to have a example project that we have in mind,
try implementing it with pipelines in BuildKite and see how it works in practice.

* [X] run longer network/protocol tests on demand

BuildKite seems to be good for general long-running tasks that you want to be able
to trigger easily, with different parameters. For example setting up a docker cluster
of jenkins nodes seems possible easily.

* [X] having a reasonable scalability path to dozens, hundreds, thousands, then even millions of nodes. with thousands, then millions of jobs; these might mean multiple tools along the way.

Very easy to spin up new agents and add them, in most platforms. Obviously not gonna handle millions of nodes since the project seems
new, but I'm sure we can work with them.




> It looks like with buildkite we still have to manually set up each repo we want CI for (same with teamcity). Are there any nice automations for setting up a 'stock CI setup for environment X'? Similar to how easy it is to get travis CI running?

@whyrusleeping, yes, it's trivial to setup our own agents with custom environments, mentioned above under my various platforms point

> The latter few are because teamcity has plugins to parse and understand test output, does buildkite have anything similar?

Seems like there is no built-in way of parsing test output and such, but if we can generate a image, we can make use of this: https://buildkite.com/docs/guides/images-in-build-output


> This is because teamcity allows us to define custom failure conditions based on both preset and user defined parameters

We can control this with environment variables and scripts in the pipelines

> It looks like buildkite supports windows, does it run scripts in cmd or git bash? (or can we specify?)

Not sure about windows specifically, I've tried with windows + ubuntu, with agents for docker-in-docker and ubuntu

---------------

Something I would love to try, is to decide a project to implement in BuildKite for example, and then see how it feels to configure and actually use it. So far, I'm quite impressed by the simplicity and configurability. I've done js-ipfs-api for testing, and it worked really well. Put some screenshots here: https://ipfs.io/ipfs/QmQDiyCBqioKskBwVWfNitonYjyoGrWm6tRxVZJUX9BoBq

**Comment by dignifiedquire** [comment]
@VictorBjelkholm I think js-ipfs-api would make a great fit. Please make sure to reenable the browser tests and see if you can make them pass on CI as we had to disable them for CI which is really not cool :/

**Comment by Kubuxu** [comment]
Failure explorer for Sharness and possibly Go tests would be extremely nice feature to have.
go-ipfs test logs are currently 2MiB in size 20k lines long, on successful builds and on failures they are even longer due to debugging logs that are being dumped. Digging through it to find a failure, even though it is close to the bottom in case of sharness as it stops running further test files on failure (IMO it would be better to run all of them even if on fails but I can see the reasoning), takes some amount of time.

**Comment by VictorBjelkholm** [comment]
@dignifiedquire Yeah, tried quickly to setup a js-ipfs-api pipeline that supports the browser tests as well as the normal steps, everything working fine!

![](https://ipfs.io/ipfs/QmdGfUc9L9tnMKrSTxsJsB2VYcLuegeHuEvNpjCBCX9RKj)

You can see the work that went into it here: https://github.com/ipfs/js-ipfs-api/compare/ci-buildkite

All in all, I'm quite happy with BuildKite and would look forward to try to integrate it with other projects as well, if we're happy with it.

**Comment by jbenet** [comment]
How can you enable viewing of this stuff without logging in? 

(it's obviously a requirement that contributors are able to view this WITHOUT having to log in -- it should have a read-only no auth mode)

**Comment by VictorBjelkholm** [comment]
> How can you enable viewing of this stuff without logging in?

@toolmantim can activate the IPFS organization as an free/open-source account, then we get public builds. Then, if we start having private builds, we have a different BuildKite organization, we just need to find a way of sharing the pipelines between them, possible via setting up a repo for composing pipelines together. 

The BuildKite IPFS Organization lives under https://buildkite.com/ipfs

> If this Buildkite org is for all third-party and public builds, I'll switch it to a free/open-source account if you like. But I'd suggest a separate org for private/sensitive/secure build pipelines, for doing all the releases etc, with a completely separate set of build agents (in their own VPC etc). ipfs-core for example. Anywho, fell free to bug me if you need a hand.

**Comment by jbenet** [comment]
@VictorBjelkholm thanks-- missed that. yeah i'd say lets make it public

**Comment by VictorBjelkholm** [comment]
So, I sent an message to the BuildKite folks about activating public builds but seems like I misunderstood the relationship between public builds and a free/open source account.

So they have a free plan for open source projects, but there is no public builds. Asked them about a timeline for public builds but they don't have any ideas about the timeline for it...

So the quest goes on...

**Comment by chriscool** [comment]
I am working on using GitLab CI to test Sharness on both Linux and Windows.

https://gitlab.com/chriscool/sharness/ is a mirror of https://github.com/chriscool/sharness/ that is automatically refreshed.

I installed a "GitLab CI Runner" on my Windows machine and I am using some shared Runners provided by http://gitlab.com to run the tests on Linux. Currently there is something wrong with the way my Windows Runner is configured so the tests on Windows fail immediately.

See: https://gitlab.com/chriscool/sharness/pipelines/4356014

But hopefully it will work soon.

**Comment by VictorBjelkholm** [comment]
@chriscool are you setting up the runner with privilege so you can do DIND (docker-in-docker) or mounting the .sock for the docker api? Seems like there is something missing with the api connection

**Comment by chriscool** [comment]
Yeah, I think I shouldn't use docker on Windows. I will try another configuration for the runner.

**Comment by dignifiedquire** [comment]
This might be intersting to look at as well https://medium.com/@hichaelmart/lambci-4c3e29d6599b#.3ydwhefgk

**Comment by chriscool** [comment]
Some GitLab related links:

- https://github.com/jerryjj/gitlab-runner-gce
- https://gitlab.com/gitlab-org/gitlab-ci-multi-runner/tree/master/docs/executors
- https://gitlab.com/gitlab-org/gitlab-ci-multi-runner/merge_requests/30
- https://gitlab.com/gitlab-org/gitlab-ci-multi-runner/merge_requests/53

From [the main GCE page](https://cloud.google.com/compute/) it looks like Google Compute Engine (GCE) has support for "Debian, CentOS, CoreOS, SUSE, Ubuntu, Red Hat, FreeBSD, or Windows 2008 R2 and 2012 R2" and "shared image from the Cloud Platform community", or our own images.

**Comment by VictorBjelkholm** [comment]
So I've sat down, looked over multiple different solutions and tried them out.
I haven't really found something that is perfect and "AH!, that's exactly what we
need" but rather I've nailed it down to a few viable options with tradeoffs, we
just have to make the right choice based on the tradeoffs and gains.

## Problems needing solution

* Slowness
* Good UI
* Configuration of server/agent/jobs
* Supports Linux/OSX/Windows/Others and different browsers
* Reliable
* Ability to do what we want AKA Supported Features
	* Should be able to run go-ipfs/js-ipfs-api integration testing
	* on demand, configurable long-running tests, performance and reliability tests
	* nightly builds
	* Github integration
	* Being able to deploy go-ipfs versions to gateways

All these are aggregated with searching through discussions on Github in IPFS
organization, from #ipfs and in person.

## Review of systems

First part of my review was to go through a list of the CI systems I could find
and aligned kind of what we want to use. You can see this list here: https://docs.google.com/spreadsheets/d/11kTloM6d1CGu_ycmP3XzpteGpJtij-h6q07anM5ZuGs/edit?usp=sharing (ignore the item on row 24, that's just my own dream CI :) )

Second part of my review was to actually try the CI's out, and here is what I found so far

### Concourse

Self-hosted. Does most things we want to do, is blazing fast and very easy to setup and maintain.
Built around BOSH but not neccessary to use. However, very young and some problems
are unsolved so far, requires a lot of workarounds for things like caching and
integration with GH for example. UI is a really nice one. Does not fit
entirely for us, yet at least.

### BuildKite

Run our own agents, perfect. UI that makes sense. Configuration of jobs that is reusable
Built for private organizations, no public UI. Would require us to design and implement
UI for that, or setup some sort of automatic inviting on PR. Too limited for
public organizations in my opinion.

### TeamCity

Self-Hosted. Giant in CI, configurable until you can't recongnize it anymore. Does pretty much
everything. Can be configured via code (Kotlin), however, no reference or docs
and very verbose. But can pretty much do everything we want, we would need to
self-host though. Also UI is not really there, being pretty bloated.

### Jenkins

Self-Hosted. Second giant in CI. Also very configurable, but the more you configure, the
more unstable it tend to become. However, covers most of our bases, and has
support for coded configuration as well. Jenkins have been used in the IPFS
project before switching to TeamCity. To note, is that Jenkins made several
large improvements since last time around in IPFS. BlueOcean is a new UI project,
making it a lot better and Jenkins 2.0 introduces native pipelines that can be coded and
reused. Another nice feature is that Jenkins can be setup to create our jobs
automatically, after pointing it to our Github organization.

### GoCD

Self-Hosted. Focus on pipelines from the ground up. Simple UI. Not very extensible, small
community. Also built for private organizations, would be require large effort
to make fit our open source nature.

### CircleCI

Hosted service. Allows us to easily setup simple jobs. Not very good for bigger
things. Easy configuration. Would require us to setup reusable jobs by ourselves.
Won't let us upgrade because everything is open source. Doesn't support windows

### TravisCI

Hosted service. We're currently using Travis, and having troubles with waiting
times and execution time, so not very fast. Simple UI though. Same applies as
with CircleCI, easy configuration for simple things but limited everywhere else.
No good for long running jobs. Can't upgrade so will seemingly forever be slow.
Does not support windows neither.

### Gitlab-CI

Gitlabs solution for doing CI. Works well, when inside the Gitlab ecosystem.
Haven't really found any good way of integrating the Gitlab-CI with Github and
it's very new, not a lot of usable thing. Independent agents are good though,
allowing us to run them ourselves. Can't find any ways of doing on-demand builds
though, or parameterized ones.

## Conclusion

In the end, we end up with two different viable options. TeamCity and Jenkins.

Pitting them against each other, they fit all the needed features, either natively
or via plugins. We'll end up hosting the infrastructure for CI ourselves, at least
for now.

### TeamCity

* TeamCity is very easy to setup new agents for.
* We're locked to having 3 agents for the free professional edition, 2500 USD for five agents...

### Jenkins

* Built-in pipelines
* Much better UI with Blue Ocean
* Better extensible and bigger community, everything exposed for plugins

------

In the end, they are mostly the same, but after going back and fourth, trying both
of them, I'd like to dive deeper into giving Jenkins a serious run.

I'm pretty confident Jenkins in the end would fit us better than TeamCity. Main
reason for switching to TeamCity from Jenkins in the beginning, was that Juan
didn't have time to maintain Jenkins and no one else could. I do have experience
with Jenkins and can take the responsibility for it.

I would like to give a larger try to Jenkins to review the new changes in Jenkins
and try to setup a cluster of agents, test-jobs for various organizations within
the IPFS project and integration test for go-ipfs/js-ipfs-api

**Comment by Kubuxu** [comment]
re CircleCI:
They allow you to pay to upgrade but only for more concurrent containers. That doesn't change that their containers are painfully slow. Comparison (my PC, CircleCI): go-ipfs coverage build 1min 14s (about 250% concurrency) vs 8 to 12 min, go-ipfs full cross-package coverage build 8min vs 40min to 1h 20min.

re TeamCity:
We have OSS license with Unlimited build agents and build configurations.

**Comment by chriscool** [comment]
re GitLab CI:
On the spreadsheet it looks like:

- it has no "Configuration in SCM", but isn't https://docs.gitlab.com/ce/ci/yaml/README.html#configuration-of-your-builds-with-gitlab-ci-yml a way to configure it in the SCM?

- it is not "Self Hosted", but it is part of GitLab and we could certainly self host it, no?

- it has no "Dashboard" but what do you mean by that exactly? Isn't something like https://gitlab.com/gitlab-org/gitlab-ce/pipelines a dashboard?

Also what do you mean by "Status Matrix" and by "Metrics"?

I am wondering how you evaluated it.

**Comment by VictorBjelkholm** [comment]
@chriscool I made the evaluation about Gitlab-CI with the assumption that we still want to host our projects on Github and not migrate them to Gitlab. "Configuration in SCM" refers to the setup of the server, which with Gitlab.com we don't control that.  If we setup our own Gitlab instance, that seems possible, but not sure we want to host Gitlab just to get the CI.

> it is not "Self Hosted", but it is part of GitLab and we could certainly self host it, no?

Sure, that is true that we could. However, we can't host just the CI, was my thinking regarding that.

> it has no "Dashboard" but what do you mean by that exactly? Isn't something like https://gitlab.com/gitlab-org/gitlab-ce/pipelines a dashboard?

That seems to be a list of the current builds and pipelines, which is good but is missing an overview of the status of the pipelines/jobs, aggregated data basically, to be able to get a quick glance if the project "healthy" or not. 

> Also what do you mean by "Status Matrix" and by "Metrics"?

Status Matrix would mean if we can build a table with current build status for different builds on different platforms. Metrics would be some sort of analytics. Gitlab includes "Cycle Analytics" which seems limited in the way that you need to use Gitlab to have use of that.

In general, those features are not deal breakers, but rather the lack of most of the other things.

**Comment by chriscool** [comment]
@VictorBjelkholm about GitLab CI you also say above "Can't find any ways of doing on-demand builds
though" but on https://gitlab.com/gitlab-org/gitlab-ce/pipelines for example there is a green "Run Pipeline" button.

About "Self Hosted", even if we would have to host GitLab, if we want GitLab CI, it seems more fair to say "Yes" than "No" (or at least something like "Yes-").

In general as with GitLab you can easily and automatically mirror any repo on GitHub or elsewhere, I don't think it is a big problem to have to use GitLab, if we want GitLab CI.

**Comment by VictorBjelkholm** [comment]
> @VictorBjelkholm about GitLab CI you also say above "Can't find any ways of doing on-demand builds
though" but on https://gitlab.com/gitlab-org/gitlab-ce/pipelines for example there is a green "Run Pipeline" button.

Yeah, I should probably have clarified the columns before sharing the spreadsheet. I'll add some definitions. Gitlab CI is marked as can do on-demand builds but no parameterized ones.

> About "Self Hosted", even if we would have to host GitLab, if we want GitLab CI, it seems more fair to say "Yes" than "No" (or at least something like "Yes-").

Sure, Gitlab we is self-hosted, my point was to evaluate the CI only. But yeah, you're right, and changed that.

> In general as with GitLab you can easily and automatically mirror any repo on GitHub or elsewhere, I don't think it is a big problem to have to use GitLab, if we want GitLab CI.

Not sure it's a great idea to replicate all repositories to Gitlab (unless we want to commit to the full Gitlab ecosystem) just to be able to use the CI, which in general doesn't fit us anyways. At least that's the idea I got from trying to get into it and trying it out.

**Comment by VictorBjelkholm** [comment]
@ipfs/go-ipfs-team @ipfs/javascript-team 
**RFC**, we need to figure out any concerns now, so we can continue moving forward.
I'll leave this for a week and if you have interest in the CI solution we
ultimately go with, please leave your feedback here.

Some things that would be good to hear from you

- [ ] Concerns about things we cannot do with Jenkins?
- [ ] Concerns in general about moving forward with Jenkins?
- [ ] Reliable/Scalable enough 
- [ ] Would using Jenkins solve our problems? Problems listed in https://github.com/ipfs/infrastructure/issues/100#issuecomment-265439892

**Comment by Kubuxu** [comment]
I don't think there is anything we could not do with Jenkins. My only thought right now is that I want us to stabilize on something.

**Comment by dignifiedquire** [comment]
@VictorBjelkholm one thing we were having issues with is having Chrome/Firefox using xvfb running on TeamCity. I would like to make sure that is addressed in whatever system we are moving to.

**Comment by Kubuxu** [comment]
There is no way around it, if it is still a problem I can promise to take a look at it.

**Comment by Kubuxu** [comment]
For future reference: https://wiki.jenkins-ci.org/display/JENKINS/TAP+Plugin

**Comment by chriscool** [comment]
@VictorBjelkholm I think triggers can be used to do parametrized builds in GitLab:

https://docs.gitlab.com/ce/ci/triggers/

especially:

https://docs.gitlab.com/ce/ci/triggers/#pass-build-variables-to-a-trigger

**Comment by lgierth** [comment]
We've had a nicely automated Jenkins instance for a while now at https://ci.ipfs.team -- the code lives at https://github.com/ipfs/jenkins
